File: _inc/_DEPRECATED_django/erp_prestech_backend/app/Http/Controllers/_helpers/error_handlers.py
from typing import Any, Union, Dict
from django.contrib import messages
from django.http import (HttpRequest, HttpResponseRedirect, 
                         HttpResponsePermanentRedirect, JsonResponse)
from django.shortcuts import redirect
from ....configs.messages_templates import get_exception_class_message
from .http import get_redirect_url
import logging
error_logger = logging.getLogger(__name__)
def default_permission_denial(request: HttpRequest,
                              err: Any,
                              ref: str = '#UNDEFINED_REFERENCE',
                              logger: logging.Logger = error_logger,
                              redirect_path: str = '/',
                              auto_redirect: bool = True,
                              json: Dict[str, Any] = None) -> Union[HttpResponseRedirect, 
                                                                   HttpResponsePermanentRedirect,
                                                                   JsonResponse,
                                                                   None]:
  from django.core.exceptions import PermissionDenied
  logger.error(f'{ref} denied permission: {err}')
  messages.error(request, get_exception_class_message(PermissionDenied, ref))
  if json:
    return JsonResponse({'error': 'Permission denied.'}, status=401)
  return redirect(get_redirect_url(request, redirect_path)) if auto_redirect else None

def default_undefined_exception(request: HttpRequest,
                                err: Any,
                                ref: str = '#UNDEFINED_REFERENCE',
                                logger: logging.Logger = error_logger,
                                redirect_path: str = '/',
                                auto_redirect: bool = True,
                                json: Dict[str, Any] = None,
                                status: int = 500) -> Union[HttpResponseRedirect,
                                                                     HttpResponsePermanentRedirect,
                                                                     JsonResponse,
                                                                     None]:
  logger.error(f'{ref} raised an undefined error: {err}')
  messages.error(request, f'An undefined error was raised: {err}')
  if json:
    return JsonResponse({'error': json}, status=status)
  return redirect(get_redirect_url(request, redirect_path)) if auto_redirect else None

File: _inc/laravel/app/Http/Controllers/Helpers/ErrorHandlers.php
<?php

namespace App\Http\Controllers\Helpers;

use App\Models\Utility;
use Illuminate\Http\{RedirectResponse, Request};
use Illuminate\Http\JsonResponse;
use Illuminate\Support\Facades\{Auth, Log, Redirect, Route};
use Symfony\Component\HttpFoundation\Response;

if (!function_exists('App\Http\Controllers\Helpers\defaultPermissionDenial')) {
	function defaultPermissionDenial(
		Request $request,
		mixed $err,
		string $ref = '#UNDEFINED_REFERENCE',
		string $redirectPath = '/',
		bool $autoRedirect = true,
		?array $json = null,
		bool $autoBack = true,
	): RedirectResponse|JsonResponse {
		Log::error("$ref denied permission", ['error' => $err]);
		session()->flash('error', 'You do not have permission to perform this action.');
		$redirectPath = $redirectPath !== '/' && str_starts_with($redirectPath, '/') && Utility::isValidRouteUrl($redirectPath) ? $redirectPath : '/';
		if (!Auth::check())
			return redirect('/login')->with('error', 'You must be logged in to access this page.');
		if (!$request->wantsJson()) {
			if ($autoBack) return redirect()->back()->with('error', 'You do not have permission to perform this action.');
			if ($autoRedirect || !$request->wantsJson()) {
				$redirectPath = $redirectPath !== '/' && str_starts_with($redirectPath, '/') && Utility::isValidRouteUrl($redirectPath) ? $redirectPath : '/';
				return !str_starts_with($redirectPath, '/') && Route::has($redirectPath)
					? redirect()->route($redirectPath) : redirect('/');
			}
		}
		return $autoRedirect
			? Redirect::to(getRedirectUrl($request, $redirectPath))
			: response()->json(['error' => 'Unauthorized', ...(is_array($json) ? $json : [])], 401);
	}
}

if (!function_exists('App\Http\Controllers\Helpers\defaultUndefinedException')) {
	function defaultUndefinedException(
		Request|\Throwable $request,
		mixed $err,
		string $ref = '# UNDEFINED_REFERENCE',
		string $redirectPath = '/',
		bool $autoRedirect = true,
		?array $json = null,
		int $status = Response::HTTP_INTERNAL_SERVER_ERROR,
		bool $autoBack = true,
	): RedirectResponse|JsonResponse {
		if ($request instanceof \Throwable) {
			Log::error("$ref raised exception: " . $request->getMessage(), ['error' => $err, 'exception_class' => get_class($request)]);
			return redirect()->back()->with('error', $request->getMessage());
		}
		// Return 404 for model-not-found exceptions instead of 500
		if ($err instanceof \Illuminate\Database\Eloquent\ModelNotFoundException) {
			Log::notice("$ref model not found", ['error' => $err->getMessage()]);
			if ($request->wantsJson() || $request->is('api/*')) {
				return response()->json(['error' => 'Resource not found', 'message' => $err->getMessage()], 404);
			}
			session()->flash('error', 'The requested resource was not found.');
			return $autoBack ? redirect()->back()->with('error', 'The requested resource was not found.')
				: redirect($redirectPath)->with('error', 'The requested resource was not found.');
		}
		$errMsg = 'Internal Server error (Http 5xx). Please notify the development team and/or your administrator.';
		Log::debug("$ref raised undefined error", ['error' => $err]);
		session()->flash('error', 'An unexpected error occurred.');
		$redirectPath = $redirectPath !== '/' && str_starts_with($redirectPath, '/') && Utility::isValidRouteUrl($redirectPath) ? $redirectPath : '/';
		if (!Auth::check())
			return redirect('/login')->with('error', $errMsg . ' Code: ' . $status);
		if (!$request->wantsJson()) {
			if ($autoBack) return redirect()->back()->with('error', $errMsg . ' Code: ' . $status);
			if ($autoRedirect || !$request->wantsJson()) {
				$redirectPath = $redirectPath !== '/' && str_starts_with($redirectPath, '/') && Utility::isValidRouteUrl($redirectPath) ? $redirectPath : '/';
				return !str_starts_with($redirectPath, '/') && Route::has($redirectPath)
					? redirect()->route($redirectPath) : redirect('/');
			}
		}
		return $autoRedirect
			? Redirect::to(getRedirectUrl($request, $redirectPath))
			: response()->json(['error' => $errMsg, ...(is_array($json) ? $json : [])], $status);
	}
}
